### ISSUE TITLE
Infinite loop with async validation (usercheck/query)

### ISSUE THREAD

**Issue by BramSikkens** [dynamoosejs/dynamoose#514]
### Summary:
Hi I'm using the validate method in my schema. The validate method uses an async function to check if a user exists based on an id. When I use the function outside the validate it works and I receive  a true/false based on the existence of the user. 
When I use it in the validate method of the schema I get an enormous loop.

#### Schema
```
var dynamoose = require('dynamoose')
var Schema = dynamoose.Schema
var ExistenceChecks = require('../Helpers/ExistenceChecks')

var UserSchema = new Schema({
  // ...

  userid: {
    type: String,
    required: true,
    hashKey: true,
    isAsync: true,
    validate: async function (v) {
      console.log('do validation')
      try {
        var result = await ExistenceChecks.checkifUserExists(v)
        // Also tried if(result === true)
        if (result.value === 'true') {
          return true
        } else {
          return false
        }
      } catch (e) {
        console.log(e)
      }

      console.log('validation done')
    }

  }
})
```
###Model 
I would save it like this
```
// Code here
  var newuser = new User({
    userid: req.body.userid, //example 1
    username: req.body.username,
    usertype: req.body.usertype,
    useremails: req.body.useremails,
    signupdate: Date.now.toString()
  })

 newuser.save(function (err) {
    if (err) {
      return console.log(err)
    }
    res.json(newuser)
  })
```

// Code here
Code to check if my user exists
```
exports.checkifUserExists = async (userid) => {
  console.log('in checkifuser exists method')
  var count = await User.queryOne('userid').eq(userid).count().exec()
  if (count > 0) {
    return true
  } else {
    return false
  }
}
```

### Current output and behavior:
in checkifuser exists method
do validation
in checkifuser exists method
do validation
in checkifuser exists method
do validation
in checkifuser exists method
do validation
in checkifuser exists method
do validation
in checkifuser exists method
do validation
in checkifuser exists method
do validation
in checkifuser exists method
do validation
in checkifuser exists method
do validation
in checkifuser exists method
do validation
in checkifuser exists method
do validation
in checkifuser exists method
do validation
in checkifuser exists method
do validation
in checkifuser exists method
do validation
.... And so on

### Expected output and behavior:
Saving of the user if the check returns true. Otherwise cancel the save 

### Environment:

Operating System: ExpressJS
Operating System Version:
Node.js version (`node -v`):v11.4.0
NPM version: (`npm -v`):6.4.1
Dynamoose version: 1.2.0
Dynamoose Plugins:

-

### Dynamoose Plugins:

- [ ] Yes, I believe that one or multiple 3rd party Dynamoose plugins are effecting this issue
- [ ] No, I believe this issue is independent of any 3rd party Dynamoose plugins I'm using
- [ ] Unknown, I'm unsure if Dynamoose plugins are effecting this issue
- [X] I am not using any Dynamoose plugins

### Type (select 1):
- [ ] Bug report
- [ ] Feature suggestion
- [X] Question
- [ ] Other suggestion
- [ ] Something not listed here

### Other:
- [X] I have read through the Dynamoose documentation before posting this issue
- [X] I have searched through the GitHub issues (including closed issues) and pull requests to ensure this issue has not already been raised before
- [X] I have searched the internet and Stack Overflow to ensure this issue hasn't been raised or answered before
- [ ] I have tested the code provided and am confident it doesn't work as intended
- [ ] I have ensured that all of my plugins that I'm using with Dynamoose are listed above
- [X] I have filled out all fields above
- [X] I am running the latest version of Dynamoose

**Comment by fishcharlie** [comment]
Async schema methods require Dynamoose version 1.3.0. You can see the release notes below.

https://github.com/dynamoosejs/dynamoose/releases/tag/v1.3.0

**Comment by BramSikkens** [comment]
Hi! 
I updated the package and now I get the error 

```
ConfigError: Missing region in config
    at Request.VALIDATE_REGION (/Users/BramSikkens/Desktop/ExpressApi/node_modules/dynamoose/node_modules/aws-sdk/lib/event_listeners.js:92:45)
    at Request.callListeners (/Users/BramSikkens/Desktop/ExpressApi/node_modules/dynamoose/node_modules/aws-sdk/lib/sequential_executor.js:106:20)
    at callNextListener (/Users/BramSikkens/Desktop/ExpressApi/node_modules/dynamoose/node_modules/aws-sdk/lib/sequential_executor.js:96:12)
    at /Users/BramSikkens/Desktop/ExpressApi/node_modules/dynamoose/node_modules/aws-sdk/lib/event_listeners.js:86:9
    at finish (/Users/BramSikkens/Desktop/ExpressApi/node_modules/dynamoose/node_modules/aws-sdk/lib/config.js:348:7)
    at /Users/BramSikkens/Desktop/ExpressApi/node_modules/dynamoose/node_modules/aws-sdk/lib/config.js:366:9
    at SharedIniFileCredentials.get (/Users/BramSikkens/Desktop/ExpressApi/node_modules/dynamoose/node_modules/aws-sdk/lib/credentials.js:126:7)
    at getAsyncCredentials (/Users/BramSikkens/Desktop/ExpressApi/node_modules/dynamoose/node_modules/aws-sdk/lib/config.js:360:24)
    at Config.getCredentials (/Users/BramSikkens/Desktop/ExpressApi/node_modules/dynamoose/node_modules/aws-sdk/lib/config.js:380:9)
    at Request.VALIDATE_CREDENTIALS (/Users/BramSikkens/Desktop/ExpressApi/node_modules/dynamoose/node_modules/aws-sdk/lib/event_listeners.js:81:26)
```

Is this a known issue after the update? I did not had this before.
I do my configuration like this: 

```
const AWS = require('aws-sdk')
AWS.config.update({
  region: 'us-east-1'
})
```

**Comment by fishcharlie** [comment]
@BramSikkens I highly doubt that is an issue with 1.3.0. Probably was an issue for you in 1.2.0 but just didn’t surface since that other problem was happening.

First ensure that configure is the very first thing that happens in your code. Before you create any schema or model or do ANYTHING related to Dynamoose.

Second change your code to the following:

```
const dynamoose = require(‘dynamoose’)
dynamoose.AWS.config.update({
  region: 'us-east-1'
})
```

It kinda looks like the backticks in the require statement are messed up since I’m on a mobile device. I would type it out manually instead of copying and pasting.

Also you can look at the documentation for more information about that.

Hope that helps.

**Comment by BramSikkens** [comment]
@fishcharlie One last question, I think the original problem was that I use, User.query in the validate method of the User schema. Causing a loop. 
Is there a way to query the User in a User schema, or should I find another way

**Comment by fishcharlie** [comment]
@BramSikkens After looking at this again I think you are 100% correct. What I said before about the Dynamoose version and configuration thing is still valid.

Sadly I don't know how to fix this off the top of my head. All of that logic is SO intertwined and makes several calls to different methods all to product the infinite loop.

It would probably require building an option on the query to skip validation, then passing that option down into the attribute creation, then handling that accordingly. Would probably need to do this for scan and get as well. And of course add documentation and tests for all of this.

There might be a more creative way to like detect if we are in an infinite loop and automatically skip validation. That might be slightly more risky and probably has a higher chance of producing unexpected results down the road.

If you'd like to work on a PR for that, that would be awesome. Otherwise I'm not sure when I'll get around to getting this done.

---

Until then, the only bandaid fix I can think of off the top of my head is to use the AWS-SDK directly. Which obviously is not ideal, but is the only solution I can think of right now until this gets done. So yeah a PR would be really appreciated.

**Comment by dobrynin** [comment]
@BramSikkens Not sure if this is still an issue for you, but is there a reason why you can't/don't want to use `condition` when invoking `save`? This would save you a database lookup. Also for this simple operation you don't need to make a model instance at all by just calling `Model.update`.  For the example you provided, this is how I would do what you are asking (assuming I'm understanding you correctly):

```
const { userid, username, usertype, useremails } = req.body;
const newuser = await User.update(
    {
         userid,
    },
    {
        username,
        usertype,
        useremails,
        signupdate: Date.now.toString()
     },
     {
          condition: 'attribute_exists(userid)',
     }
);
```

I used some ES6+ syntax, hope that's alright.

The downside of this is that you have to specify it with every invocation. But do you really want to limit your model from not being able to create new records?

Another thought:
```
    const { userid, username, usertype, useremails } = req.body;
    const user = await User.get(userid);
    if (user) {
        user.username = username;
        user.usertype = usertype;
        user.useremails = useremails;
        user.signupdate = Date.now().toString();
        await user.save();
        res.json(user);
    } else {
        console.log('No user found');
    }
```

If neither of these work for you, perhaps the solution is to provide a top-level `condition` option on a model that will be included with all `save` and `update` actions. Or maybe `validate` is the way to go. Would like to hear some feedback on this.

**Comment by dobrynin** [comment]
I'm also a bit confused because it seems to me like you are trying to create a "new user" but are also trying to disallow saving a model unless the user already exists.

**Comment by fishcharlie** [comment]
I made a note in the documentation about this (#561). Closing for now. If anyone has a creative way to handle this better I'd love to see a PR! 😃

**Comment by purefan** [comment]
Somewhat related to this I just ran into a situation where I wanted to check the existence of a `role` when creating a `user`. I was able to validate that a `role` exists in another table by moving the `const db = require('/path/to/models')` inside the validate function in the user Schema. 

I think that when I had it at the top of the file it just wasn't initialized, no error was thrown but it was just an empty object.
